client/src/lib/paymentCalc.ts

```diff
@@ -1,58 +1,104 @@
+import { getAxiomV2Abi, getOpStackGasPriceOracleAbi, getOpStackL1BlockAttributesAbi } from "./abi";
+import { isArbitrumChain, isMainnetChain, isOpStackChain, isScrollChain } from "./chain";
+import { publicActionsL2 } from 'viem/op-stack';
 
 export async function calculatePayment(
-  axiomV2QueryAddr: string,
+  chainId: string,
   publicClient: PublicClient,
-  options?: AxiomV2QueryOptions
+  options?: AxiomV2ClientOptions,
 ): Promise<bigint> {
+  const axiomV2QueryAddr = options?.overrides?.queryAddress ?? getAxiomV2QueryAddress(chainId);
+
+  // Convert callback gas limit to wei
+  const callbackGasLimit = BigInt(options?.callbackGasLimit ?? ClientConstants.DEFAULT_CALLBACK_GAS_LIMIT);
+
+  // Get maxFeePerGas
+  const maxFeePerGas = BigInt(options?.maxFeePerGas ?? ClientConstants.DEFAULT_MAX_FEE_PER_GAS_WEI);
+
   // Get proofVerificationGas from contract
-  let proofVerificationGas;
-  try {
-    proofVerificationGas = await publicClient.readContract({
-      address: axiomV2QueryAddr,
-      abi: getAxiomV2Abi(AbiType.Query),
-      functionName: "proofVerificationGas",
-      args: [],
-    }) as bigint; // in gas units
-  } catch (e) {
-    console.log(`Unable to read proofVerificationGas from contract`);
-  }
-  proofVerificationGas = BigInt(proofVerificationGas ?? 0);
-  if (proofVerificationGas === 0n) {
-    proofVerificationGas = ClientConstants.FALLBACK_PROOF_VERIFICATION_GAS;
-  }
+  const proofVerificationGas = await getAxiomV2QueryValue(
+    publicClient,
+    axiomV2QueryAddr,
+    "proofVerificationGas",
+    [],
+    ClientConstants.FALLBACK_PROOF_VERIFICATION_GAS
+  );
 
-  // Get axiomQueryFee from contract
-  let axiomQueryFee;
-  try {
-    axiomQueryFee = await publicClient.readContract({
-      address: axiomV2QueryAddr,
-      abi: getAxiomV2Abi(AbiType.Query),
-      functionName: "axiomQueryFee",
-      args: [],
-    }) as bigint; // in wei
-  } catch (e) {
-    console.log(`Unable to read axiomQueryFee from contract`);
+  // Get the projected callback cost; mainnet will return 0
+  const projectedCallbackCost = await getProjectedCallbackCost(chainId, publicClient, maxFeePerGas, callbackGasLimit, proofVerificationGas);
+
+  // Get overrideAxiomQueryFee from either equation or options
+  let overrideAxiomQueryFee: bigint;
+  if (options?.overrideAxiomQueryFee !== undefined) {
+    overrideAxiomQueryFee = BigInt(options.overrideAxiomQueryFee);
+  } else {
+    // overrideAxiomQueryFee = AXIOM_QUERY_FEE + projectedCallbackCost - maxFeePerGas * (callbackGasLimit + proofVerificationGas)
+    overrideAxiomQueryFee = ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI + 
+      projectedCallbackCost - maxFeePerGas * (callbackGasLimit + proofVerificationGas);
   }
-  axiomQueryFee = BigInt(axiomQueryFee ?? 0);
-  if (axiomQueryFee === 0n) {
-    axiomQueryFee = ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI;
+
+  // Get axiomQueryFee from contract if on ethereum, otherwise use L2 equation
```

Comment:
Can we re-arrange the logic here to more explicitly branch on the chainId?  It's hard to follow the flow in the L1 and L2 case right now because they are mixed